[clang] CTAD: fix the aggregate deduction guide for alias templates.

hokein opened the pull request:
For alias templates, our current way of constructing their aggregate deduction guides deviates from the standard approach. We should align it with how we handle implicit deduction guides.

This patch has a refactoring change which pulls the construction logic out from `DeclareImplicitDeductionGuidesForTypeAlia` and reusing it for building aggregate deduction guides.

erichkeane commented on clang/test/SemaTemplate/deduction-guide.cpp:
```diff
@@ -261,6 +261,13 @@ AG ag = {1};
+using BG = G<int>;
+BG bg = {1.0};
+// CHECK-LABEL: Dumping <deduction guide for BG>
+// CHECK: FunctionTemplateDecl
```
Could these be `CHECK-NEXT` ?

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2803,7 +2803,207 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+// Build deduction guides for a type alias template from the given underlying
+// deduction guide F.
+FunctionTemplateDecl *
+BuildDeductionGuideForTypeAlias(Sema &SemaRef,
```
Best I can tell this is mostly a Copy/paste from the other two locations?  Can you point out for me what are the meaningful different changes?

sdkrystian replied:
(if possible, please put the non-functional changes into their own commit for review purposes)

hokein replied:
We just move the implementation from `DeclareImplicitDeductionGuidesForTypeAlias ` to a dedicated function, so that we can reuse it in the `DeclareAggregateDeductionGuideForTypeAlias`.

The functional change of this commit is in `DeclareAggregateDeductionGuideForTypeAlias`, the previous implementation is incorrect, and we should use the one in `DeclareImplicitDeductionGuidesForTypeAlias`.

hokein replied:
I have restructured commits in this PR. It contains two commits: one for the NFC refactoring change, the other one is for the actual fix. I hope it would make the review easier.

cor3ntin replied:
to clarify, in the actual fix you removed a whole bunch of code?
https://github.com/llvm/llvm-project/pull/90894/commits/621899224aa7d59ab70675f6ff32b298b06a515a

erichkeane commented:
Also needs release note.

cor3ntin commented:
> Also needs release note.

that's a clang 19 feature fix afaict, so nope

shafik commented on clang/test/SemaTemplate/deduction-guide.cpp:
```diff
@@ -1,4 +1,4 @@
-// RUN: %clang_cc1 -std=c++2a -verify -ast-dump -ast-dump-decl-types -ast-dump-filter "deduction guide" %s | FileCheck %s --strict-whitespace
+// RUN: %clang_cc1 -std=c++2a -verify -ast-dump -ast-dump-decl-types -Wno-c++11-narrowing -ast-dump-filter "deduction guide" %s | FileCheck %s --strict-whitespace
```
Is there a way to verify this fix w/o using a narrowing conversion?

hokein committed:
Refactor: Extract the core deduction-guide construction implementation from DeclareImplicitDeductionGuidesForTypeAlias

We move the core implementation to a dedicate function, so that it can
be reused in other places.

hokein committed:
[clang] CTAD: fix the aggregate deduction guide for alias templates.

For alias templates, the way we construct their aggregate deduction guides is
not following the standard way. We should do the same thing as we do for
implicit deduction guides.

hokein commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2803,7 +2803,207 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+    Args.addOuterTemplateArguments(TransformedDeducedAliasArgs);
+    NamedDecl *NewParam = transformTemplateParameter(
+        SemaRef, AliasTemplate->getDeclContext(), TP, Args,
+        /*NewIndex*/ FPrimeTemplateParams.size());
```
Done.

erichkeane reviewed (approved):
I think I'm ok with this, please give others a day or two to sign off as well.

mizvekov commented on clang/test/SemaTemplate/deduction-guide.cpp:
```diff
@@ -261,6 +261,13 @@ AG ag = {1};
+BG bg(1.0);
+// CHECK-LABEL: Dumping <deduction guide for BG>
+// CHECK:      FunctionTemplateDecl {{.*}} implicit <deduction guide for BG>
+// CHECK: |-CXXDeductionGuideDecl {{.*}} 'auto (int) -> G<int>' aggregate
```
Nit: Since this isn't matching the structure of the AST anyway, it would probably be a good idea to avoid the possibility of unrelated test churn.

```suggestion
// CHECK: CXXDeductionGuideDecl {{.*}} 'auto (int) -> G<int>' aggregate
```

mizvekov replied:
The line is a CHECK, not a CHECK-NEXT, so it could appear any number of lines afterwards.
So this means it could be a child of something else that appears after it.

The CHECK matcher matches any number of characters before the first character specified,
So it would match something like `| | |-CXXDeductionGuideDecl {{.*}} 'auto (int) -> G<int>' aggregate`
Ie, it could be deeply nested into something else.

Instead, I think it just demonstrates:
* It is a child of something.
* It's neither the first nor the last child of that thing.

hokein replied:
> The line is a CHECK, not a CHECK-NEXT, so it could appear any number of lines afterwards.

Yeah, ideally it should be `CHECK-NEXT`, but we can't use it because of the ambiguity in the content of the above line (there are two other deduction guides emitting the same content), so I end up using the `CHECK: |-...`, it at least demonstrates that it is a child of something.

mizvekov replied:
But that it is a child of something is trivial, because all declarations are children of something, except a translation unit declaration.

But the fact that this line test that this thing is neither the first nor the last child of that thing, means that it is quite likely to break with unrelated test churn.

mizvekov reviewed (approved):
This looks like a straightforward fix to me as well

LGTM

sdkrystian reviewed (approved):
LGTM

hokein committed:
Address comments.
